pkg/volume/gce_pd/attacher_test.go

```diff
@@ -0,0 +1,358 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce_pd
+
+import (
+	"errors"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetest "k8s.io/kubernetes/pkg/volume/testing"
+
+	"github.com/golang/glog"
+)
+
+func TestGetDeviceName_Volume(t *testing.T) {
+	plugin := newPlugin()
+	name := "my-pd-volume"
+	spec := createVSpec(name, false)
+
+	deviceName, err := plugin.GetDeviceName(spec)
+	if err != nil {
+		t.Errorf("GetDeviceName error: %v", err)
+	}
+	if deviceName != name {
+		t.Errorf("GetDeviceName error: expected %s, got %s", name, deviceName)
+	}
+}
+
+func TestGetDeviceName_PersistentVolume(t *testing.T) {
+	plugin := newPlugin()
+	name := "my-pd-pv"
+	spec := createPVSpec(name, true)
+
+	deviceName, err := plugin.GetDeviceName(spec)
+	if err != nil {
+		t.Errorf("GetDeviceName error: %v", err)
+	}
+	if deviceName != name {
+		t.Errorf("GetDeviceName error: expected %s, got %s", name, deviceName)
+	}
+}
+
+func TestAttach_Positive(t *testing.T) {
+	// Attach the disk on the first attempt
+	diskName := "disk"
+	instanceID := "instance"
+	readOnly := false
+	attachCalls := []attachCall{
+		{diskName, instanceID, readOnly, nil},
+	}
+	isAttachedCalls := []diskIsAttachedCall{
+		{diskName, instanceID, false, nil},
+	}
+	gce := newFakeGCE(t, attachCalls, nil, isAttachedCalls)
+	attacher := newAttacher(gce)
+
+	spec := createVSpec(diskName, readOnly)
+
+	err := attacher.Attach(spec, instanceID)
+	if err != nil {
+		t.Errorf("Attach failed: %v", err)
+	}
+}
+
+func TestAttach_Negative_Attach(t *testing.T) {
+	// Attach attempt fails
+	diskName := "disk"
+	instanceID := "instance"
+	readOnly := false
+	attachCalls := []attachCall{
+		{diskName, instanceID, readOnly, errors.New("Fake attach error")},
+	}
+	isAttachedCalls := []diskIsAttachedCall{
+		{diskName, instanceID, false, nil},
+	}
+	gce := newFakeGCE(t, attachCalls, nil, isAttachedCalls)
+	attacher := newAttacher(gce)
+
+	spec := createVSpec(diskName, readOnly)
+
+	err := attacher.Attach(spec, instanceID)
+	if err == nil {
+		t.Errorf("Attach failure expected, got nil")
+	}
+}
+
+func TestAttach_Positive_DiskIsAttached(t *testing.T) {
+	// DiskIsAttached query fails, but the plugin still tries to attach the
```

Comment:
Why not do table tests for these?  They should be doable as a table test pretty easily for less lines.